org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/notifications/NotificationsManager.java

```diff
@@ -0,0 +1,162 @@
+  }
+
+  // only for testing
+  public NotificationsManager(Subscriber subscriber, SonarLintProjectConfigurationReader configReader) {
+    this.subscriber = subscriber;
+    this.configReader = configReader;
+  }
+
+  public synchronized void subscribe(ISonarLintProject project, SonarQubeNotificationListener listener) {
+    SonarLintProjectConfiguration config = configReader.apply(project);
+
+    String projectKey = config.getProjectKey();
+    if (projectKey == null) {
+      return;
+    }
+
+    Set<String> moduleKeys = subscribers.get(projectKey);
+    if (moduleKeys == null) {
+      if (!subscriber.subscribe(project, config, listener)) {
+        return;
+      }
+      moduleKeys = new HashSet<>();
+      subscribers.put(projectKey, moduleKeys);
+      listeners.put(projectKey, listener);
+    }
+
+    String moduleKey = config.getModuleKey();
+    moduleKeys.add(moduleKey);
+  }
+
+  /**
+   * Read and save directly from the mutable object.
+   * Any changes in the project settings will affect the next request.
+   */
+  // visible for testing
+  public static class ProjectNotificationTime implements LastNotificationTime {
+
+    private final NotificationsTracker tracker;
+
+    // visible for testing
+    public ProjectNotificationTime(NotificationsTracker tracker) {
+      this.tracker = tracker;
+    }
+
+    public ProjectNotificationTime(ISonarLintProject project) {
+      this(SonarLintCorePlugin.getOrCreateNotificationsTracker(project));
+    }
+
+    @Override
+    public ZonedDateTime get() {
+      ZonedDateTime lastEventPolling = tracker.getLastEventPolling();
+      if (lastEventPolling == null) {
+        lastEventPolling = ZonedDateTime.now();
+        tracker.setLastEventPolling(lastEventPolling);
+      }
+      return lastEventPolling;
+    }
+
+    @Override
+    public void set(ZonedDateTime dateTime) {
+      // this could be false if the settings changed between the read and write
+      if (dateTime.isAfter(tracker.getLastEventPolling())) {
+        tracker.setLastEventPolling(dateTime);
+      }
+    }
+  }
+
+  public synchronized void unsubscribe(ISonarLintProject project) {
+    SonarLintProjectConfiguration config = configReader.apply(project);
+
+    String projectKey = config.getProjectKey();
+    Set<String> moduleKeys = subscribers.get(projectKey);
+    if (moduleKeys == null) {
+      return;
+    }
+
+    String moduleKey = config.getModuleKey();
+    moduleKeys.remove(moduleKey);
+
+    if (moduleKeys.isEmpty()) {
+      subscribers.remove(projectKey);
+      subscriber.unsubscribe(listeners.get(projectKey));
+    }
+  }
+
+  // visible for testing
+  public static class Subscriber {
+    // visible for testing
+    public boolean subscribe(ISonarLintProject project, SonarLintProjectConfiguration config, SonarQubeNotificationListener listener) {
+      Server server = (Server) SonarLintCorePlugin.getServersManager().getServer(config.getServerId());
+      if (!server.areNotificationsEnabled()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "server" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)